Adding GUI for gamma-launcher

sobkas opened the pull request:
Hi,
Have created simple GUI for gamma-launcher to make it easier to use.

sobkas committed:
Add GUI, add GTK4 dependecies

sobkas committed:
Add support for displaying console output
Use Gtk.ApplicationWindow class more directly

sobkas committed:
Add support to disable/enable mods

sobkas committed:
Make file entry fields persistant between tabs

sobkas committed:
Debug option to disable stdout,stderr redirect

sobkas committed:
Add cache dir to Partial Gamma Setup

sobkas committed:
For older versions of Python there is need to call GLib.threads_init()

PyGIDeprecationWarning: Since version 3.11, calling threads_init is no longer needed. See:
https://pygobject.gnome.org/guide/threading.html#threads-faq

sobkas commented:
>Execution output widget cant be resized to take the whole window's space left
It resizes automatically as text fills it, it can be scrolled

>Stored config args can be parsed to prefill widgets
Don't know about it yet, had problems in the past with config file for gamma-launcher

>A folder selection dialog can be used for all directories widgets
First a working prototype, then adding some ease of use improvments

>7z output is redirected to stdout instead of output widget. (use fork instead of thread ?)
I think it can be fixed in gamma-launcher itself

>Useful ? I mean ... I don't think a lot of people will use this command through the GUI
I was thinking about removing it, only added it by accident because I mistook it for full-install

>Setting GAMMA dir and clicking on Load have 0 effect
While Executing one command(button set to Executing, need better indicator to show that something is going) every other command ignores buttons
So if other action is going (Button for it is set to "Executing" no other action can take place except Mod Chooser
In case of mod chooser I should add error if it can't find modlist.txt

>Can't run after filling up Anomaly dir / Gamma dir / CACHE dir (custom option are optional, so empty). Just failing, no error in the output widget (even with --debug option)
Do you see :"Redirecting stdout and stderr" ?
Does last commit fixes it
Does button text changes into Executing?
If it's Executing it's going but just doesn't print anything
Because I have:
![Zrzut ekranu z 2025-06-26 23-34-30](https://github.com/user-attachments/assets/cfd0ef25-9480-4c76-8bb7-d86b22ef5af4)

sobkas committed:
If Gamma Dir ended without '/' it ignored that dir

sobkas commented:
Ok, my bad in mod chooser, if directory didn't have / at the end, it didn't detect modlist.txt

sobkas committed:
Fix Module Chooser success/fail modes

Mord3rca commented on launcher/gui.py:
```diff
@@ -0,0 +1,322 @@
+
+    def load_mods(self, button):
+        self.file_name = self.entries['Mod Chooser'][0].get_chars(0, -1)
+        self.file_name += '/profiles/G.A.M.M.A/modlist.txt'
```
Use `pathlib.Path` to avoid / issue ... also, it will work on Windows too.

`self.file_name = Path(self.file_name) / 'profile' / 'G.A.M.M.A' / 'modlist.txt'`

Mord3rca committed:
Installing **PyGObject** dependencies for PR checks

According to https://pygobject.gnome.org/getting_started.html#ubuntu-logo-ubuntu-debian-logo-debian

Omitting **gir1.2-gtk-4.0** since running the GUI is not planned.

Mord3rca commented:
Can you cherry-pick this commit https://github.com/Mord3rca/gamma-launcher/commit/0237af2fd83b31f1fe1f46b2024016fe916f6345 to fix pycairo install for automated linting ?

If you don't know how to do this, here is all the git commands:
```sh
git remote add upstream https://github.com/Mord3rca/gamma-launcher
git fetch upstream
git cherry-pick 0237af2fd83b31f1fe1f46b2024016fe916f6345
```

sobkas committed:
Switch to pathlib

Mord3rca commented:
Mod chooser tab is now functional :)

Full Gamma Setup / Partial Gamma Setup is still KO
![Screenshot_2025-06-27_20-18-26](https://github.com/user-attachments/assets/efcd0a04-7f33-4c66-95ba-b0b1ce72fdcd)


And output widget still don't extended correctly on my side
![Screenshot_2025-06-27_20-06-00](https://github.com/user-attachments/assets/e16565c8-67f1-4896-9ebd-79d12ab7ed2d)


Note: I'm using Python 3.12 and here is my pip env list:
```sh
Package            Version
------------------ ---------
beautifulsoup4     4.12.3
Brotli             1.1.0
certifi            2024.8.30
charset-normalizer 3.4.0
click              7.1.2
cloudscraper       1.2.71
gamma-launcher     2.3
gitdb              4.0.11
GitPython          3.1.43
idna               3.10
inflate64          1.0.0
multivolumefile    0.2.3
pip                25.1.1
platformdirs       4.3.6
psutil             6.1.0
py7zr              0.22.0
pybcj              1.0.2
pycairo            1.28.0
pycryptodomex      3.21.0
PyGObject          3.52.3
pyparsing          3.2.0
pyppmd             1.1.0
pyzstd             0.16.2
requests           2.32.3
requests-toolbelt  1.0.0
ruff               0.8.4
smmap              5.0.1
soupsieve          2.6
tenacity           9.0.0
texttable          1.7.0
tqdm               4.67.0
unrar              0.4
urllib3            2.2.3
```

And here are the GTK and GTK related libs on my system:
```sh
dev-libs/glib-2.82.5
dev-libs/gobject-introspection-1.82.0-r2
x11-libs/cairo-1.18.4-r1
x11-libs/gdk-pixbuf-2.42.12
x11-libs/gtk+-2.24.33-r3
x11-libs/gtk+-3.24.49-r1
x11-libs/pango-1.52.2
```

sobkas committed:
Provide info about Exception

sobkas committed:
Fix some flake8 errors, provide more info for Mod Chooser

sobkas commented:
Now It should print error properly, had to change max-line-length to 180

sobkas committed:
Remove Partial Gamma Setup, file inputs are saved in config file, supress E402

Mord3rca commented:
> had to change max-line-length to 180 in .flake8

For which line ? [This one](https://github.com/sobkas/gamma-launcher/blob/master/launcher/gui.py#L125) ?
You can rewrite it this way:
```python
self.generic_tab(
  'Full Gamma Setup',
  ['Anomaly Dir', 'Gamma Dir', 'Custom Gamma Definition', 'Custom Gamma Repository', 'Cache Dir'],
  self.full_install
)
```

You can do the same thing for all Thread lines. No need to globally allow lines >180 char.


On another note, Full Install / Gamma setup are now working :+1:

Mord3rca committed:
Making gui installation optional

Mord3rca commented:
Can you also cherry-pick commit fd1563ac491c7184ab99504d05d438b5d83a8cae ?

This will make your modification optional in **pyproject.toml** and document it.

To do so, here the git commands (assuming upstream remote is already there):
```sh
git fetch upstream
git cherry-pick fd1563ac491c7184ab99504d05d438b5d83a8cae
```

sobkas committed:
Allow use of File Dialog for choosing folders

sobkas commented:
Ok added option to use File dialog to chose folders

sobkas commented:
Ok fixed look of folder icon, now next to entry field

sobkas committed:
Make debian package, move to unrardll, add icon and desktop files.

sobkas committed:
Show errors during debug

sobkas committed:
Clean Gamma dir when updating GAMMA definition

sobkas committed:
Remove old files from GAMMA usvfs workaround dir

sobkas committed:
Properly display separators

sobkas committed:
Preserve config files

sobkas committed:
Expand user for mod list filename

sobkas committed:
Fix display of mod names

sobkas committed:
Fix flak8 errors

sobkas committed:
Push new changelog

Mord3rca commented on launcher/commands/usvfs.py:
```diff
@@ -34,8 +34,16 @@ def run(self, args) -> None:
         install_dir = Path(args.final).expanduser()
-
-        install_dir.mkdir(parents=True)
+        if install_dir.exists():
```
Why not doing this instead ?
* Reading *user.ltx*
* Deleting install_dir (with rmtree)
* mkdir()

Mord3rca commented on launcher/commands/usvfs.py:
```diff
@@ -50,3 +58,5 @@ def run(self, args) -> None:
         print('Reapplying Anomaly binary dir to install directory')
         copytree(anomaly_dir / 'bin', install_dir / 'bin', dirs_exist_ok=True)
+        if Path(install_dir / 'appdata' / 'user.ltx~').exists():
+            copy2(install_dir / 'appdata' / 'user.ltx~', install_dir / 'appdata' / 'user.ltx')
```
If following the last comment of this file, just rewrite user.ltx with the previous content

```python
if content_user_ltx:
    Path(...).write_text(content_user_ltx)
```

Mord3rca reviewed (changes requested):
Some action are required:
* Remove Debian packaging
* Remove modification on *archive.py*
* Move ModChooser logic to a command
* *.flake8* should not be modified

Also, GUI is running but I'm not a GUI expert + I don't plan to use it.
Please see with some users to test it first and put their feedback on this PR

Mord3rca commented:
Also f51c7d52b81e9e8c986124e8734b92329821c7b3 can be dropped since PR check wont install the GUI
